### PR TITLE
type annotations for dataloader, dataset, sampler

### LINKED ISSUE #38913: Typing Error in torch.utils.data.DataLoader and Dataset
## 🐛 Bug

In [dataloader.py#60](https://github.com/pytorch/pytorch/blob/d035d05080729c30636ff30fcc068de3c7e9badd/torch/utils/data/dataloader.py#L60), `DataLoader` does not inherit `typing.Generic`, but in [dataloader.pyi#15](https://github.com/pytorch/pytorch/blob/d035d05080729c30636ff30fcc068de3c7e9badd/torch/utils/data/dataloader.pyi#L15) type inferace of `DataLoader` inherits `typing.Generic`.

`Dataset` is also in same situation. See also [dataset.py#L8](https://github.com/pytorch/pytorch/blob/d035d05080729c30636ff30fcc068de3c7e9badd/torch/utils/data/dataset.py#L8), [dataset.pyi#L6](https://github.com/pytorch/pytorch/blob/d035d05080729c30636ff30fcc068de3c7e9badd/torch/utils/data/dataset.pyi#L6).

## To Reproduce

```python
>>> import torch
>>> from typing import List
>>> from torch.utils.data import Dataset, DataLoader
>>> class SomeDatasetClass(Dataset[List[torch.Tensor]]):
...     ...
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'type' object is not subscriptable
```

```python
>>> import torch
>>> from typing import List
>>> from torch.utils.data import Dataset, DataLoader
>>> class SomeDatasetClass(Dataset):
...     ...
... 
>>> def _create_dataloader(is_train: bool) -> DataLoader[List[torch.Tensor]]:
...     ...
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'type' object is not subscriptable
```

## Expected behavior

`DataLoader` and `Dataset` should inherit `typing.Generic`, so above code should run without any exceptions.

```python
>>> from typing import Generic, TypeVar
>>> T = TypeVar('T')
>>> class SomeGenericClass(Generic[T]):
...     ...
... 
>>> def create_some_generic_class() -> SomeGenericClass[bool]:
...     ...
...
>>>
```

Like above code snippet.

## Environment

 - PyTorch Version (e.g., 1.0): 1.5.0
 - Python version: Python 3.7.5

## Additional context

I think this issue is related to #31765

cc @ezyang @gchanan @zou3519

### PR DESCRIPTION
**Pull request by Baranowski**
Fixes #38913

### CODE REVIEW

File: torch/utils/data/sampler.py
Comment (zou3519) [inline torch/utils/data/sampler.py:55]: According to the note at https://github.com/pytorch/pytorch/blob/2b29feace4d4711190ad0a5012ba517f45e0e336/torch/utils/data/sampler.py#L23-L48 , we should not be providing a default implementation of `__len__`. Do we need SizedSampler for all the types to check out? Scrolling through this PR it looks like SizedSampler is only used to construct the other Samplers

File: torch/utils/data/sampler.py
Comment (zou3519) [inline torch/utils/data/sampler.py:231]: What's up with the `# type: ignore` ? Might be good to include the reason as a code comment as well

File: torch/utils/data/dataloader.py
Comment (zou3519) [inline torch/utils/data/dataloader.py:137]: `batch_size` was original `int`, why did this have to change?

File: torch/utils/data/dataloader.py
Comment (zou3519) [inline torch/utils/data/dataloader.py:369]: This one doesn't need to be quoted, right?

File: torch/utils/data/dataloader.py
Comment (zou3519) [inline torch/utils/data/dataloader.py:233]: Just checking -- this is a pre-existing problem, right? I see that RandomSampler before this PR takes `Sized`, and Dataset before this PR is a `Generic[T_co]`

**Review by zou3519 — COMMENTED** [review]
I had some questions/suggestions. The PR generally looks fine to me otherwise.

I'm not very familiar with how type hints for these files are tested in our CI. Is the `test_typing` in `test_dataloader.py` sufficient to check everything, or are there other tests somewhere?

File: torch/utils/data/dataloader.py
Comment (Baranowski) [reply torch/utils/data/dataloader.py:137]: There is a condition in the body of this method which checks whether `batch_size` is `None`, so `None` is an allowed value.

Separately from this, this argument is also used as a local variable in the method and sometimes `None` is assigned to it.

File: torch/utils/data/dataloader.py
Comment (Baranowski) [reply torch/utils/data/dataloader.py:369]: I'm getting `NameError: name '_BaseDataLoaderIter' is not defined` if I remove the quotes (mypy 0.770)

File: torch/utils/data/dataloader.py
Comment (Baranowski) [reply torch/utils/data/dataloader.py:233]: That's correct

File: torch/utils/data/sampler.py
Comment (Baranowski) [reply torch/utils/data/sampler.py:234]: I have just expanded on the comment two lines above.

**Comment by Baranowski** [comment]
Thanks for the review @zou3519. Re: tests, there is also `test/test_type_hints.py` which runs `mypy`.

File: torch/utils/data/dataloader.py
Comment (zou3519) [reply torch/utils/data/dataloader.py:370]: Sad, mypy limitations :/

**Review by zou3519 — APPROVED** [review]
I'm not a big fan of the `# type: ignore`s, but it doesn't look like we can do anything about it and this PR also makes the current state better than before so let's ship it

**Review by rgommers — COMMENTED** [review]
LGTM too, thanks @Baranowski. 

One thing I noticed is that this is still present in `mypy.ini`:
```
[mypy-torch.utils.data.dataset]
ignore_errors = True
```
I suspect you've run `mypy` directly on the file to test your changes, in which case the ignore doesn't matter. In CI the ignore does matter though, so I'd recommend to remove it in this PR (or as a follow up) if tests `test_type_hints.py` passes locally.

**Comment by Baranowski** [comment]
I was relying on test_type_hints.py, so I will need to fix up dataset.py as well

File: torch/utils/data/dataloader.py
Comment (rgommers) [inline torch/utils/data/dataloader.py:301]: This is no longer necessary - mypy 0.770 is the minimum requirement.

File: torch/utils/data/dataset.py
Comment (rgommers) [inline torch/utils/data/dataset.py:167]: This changes the type of `self.tensors`, which is a public attribute, from `tuple` to `list` and does extra work - seems undesirable. Why was this change needed?

**Review by rgommers — APPROVED** [review]
LGTM modulo the couple of minor open comments.

The PR is still draft, I assume it's not anymore - looks about ready.

File: torch/utils/data/dataloader.py
Comment (Baranowski) [reply torch/utils/data/dataloader.py:301]: If I remove the quotes, I'm still getting `NameError: name '_BaseDataLoaderIter' is not defined` from mypy 0.770. I have removed the part of the comment about newer versions of mypy handling this

Comment (rgommers) [reply torch/utils/data/dataloader.py]: Sounds good, thanks

**Comment by Baranowski** [comment]
The CI failure looks unrelated so this PR should be ready now.

**Review by zou3519 — APPROVED** [review]
lgtm as well

**Review by zou3519 — COMMENTED** [review]
There are some other internal failures that I am reading, will report back

File: torch/utils/data/dataset.py
Comment (Baranowski) [reply torch/utils/data/dataset.py]: Done

File: torch/utils/data/dataloader.py
Comment (zou3519) [inline torch/utils/data/dataloader.py:133]: Is it possible to list `sampler: Sampler` here? Some internal code attempts to access `dataloader.sampler` and the type appears was inferred to be Optional[Sampler]. However, after `__init__`, we're sure that sampler is a Sampler

**Review by zou3519 — COMMENTED** [review]
I think this is the last external change we need. I have some code ready to change some (previously incorrectly typed) internal code snippets that are using Dataloader / Dataset

File: torch/utils/data/dataloader.py
Comment (zou3519) [reply torch/utils/data/dataloader.py:133]: (I verified that adding `sampler: Sampler` makes the problem go away, but I'm not sure if there was a reason why it wasn't here)

**Comment by zou3519** [comment]
@Baranowski, @SsnL, how do you feel about adding an abstract `__len__` method to Sampler? https://github.com/pytorch/pytorch/blob/e440c370c5ea47e5e4345788320391a51154b50f/torch/utils/data/sampler.py#L23

The rationale behind this is that I'm sure there is a lot of user code out there that checks `len(loader.sampler)` and it feels wrong for that to not type-check if most of the Samplers include with PyTorch support `__len__`. Making Sampler inherit from `typing.Sized` gives Sampler a `__len__` function that returns a TypeError.

@SsnL I read https://github.com/pytorch/pytorch/blob/e440c370c5ea47e5e4345788320391a51154b50f/torch/utils/data/sampler.py#L46, and I think if we verify that for all versions of Python that we support, if the following gives something sensible, then it should be an acceptable solution. 
```
class Foo(typing.Sized):
    pass
len(Foo())
```

**Comment by Baranowski** [comment]
@zou3519 I don't have enough experience with Python to have a valuable opinion. I'm happy to do whatever you guys think is best.

**Comment by SsnL** [comment]
@zou3519 However the sampler doesn't necessarily have to be `Sized` (i.e., have a working `__len__` though). Really it is just required to be an `Iterable`. 

If we make `Sampler` inherit from `Sized` and `Iterable` and not implement `__len__` then `abc` complains:
```py

In [4]: class A(typing.Sized, typing.Iterable):
   ...:     def __iter__(self): yield from [1,2,3]
   ...:

In [5]: A()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-6234893e030b> in <module>
----> 1 A()

~/miniconda3/lib/python3.7/typing.py in __new__(cls, *args, **kwds)
    816             obj = super().__new__(cls)
    817         else:
--> 818             obj = super().__new__(cls, *args, **kwds)
    819         return obj
    820

TypeError: Can't instantiate abstract class A with abstract methods __len__
```

I don't know if we should be adding a `__len__` that just raises `TypeError` to resolve this... Maybe we should?

**Comment by zou3519** [comment]
@SsnL thanks for the context and testing that. I'm not sure if we should raise a TypeError either. It sounds like we should keep things as is-right now (Sampler without `__len__` method), and mark this PR as potentially bc-breaking (it can break type-checked user code that tries to access `dataloader.sampler`; the workaround would be to add a `# type: ignore` or to use `typing.cast` to cast the sampler to the expected sampler type).

**Comment by Baranowski** [comment]
@zou3519 @SsnL so as I understand, there is nothing more for me to do here? Is hit ready to import and merge?

**Comment by zou3519** [comment]
> so as I understand, there is nothing more for me to do here? Is this ready to import and merge?

Yeah that's correct. I'm working on the import and merge

**Comment by zou3519** [comment]
@Baranowski actually, could you please rebase this PR and resolve the merge conflict? I think it should just be accepting the changes to `tools/pyi/gen_pyi.py`

@Baranowski to provide some more transparency, we have some internal code that previously wasn't type checked but now is type checked as a result of adding the annotations. I've been working on correcting the annotations for that code and am done with that, so this should be good to go very soon after we resolve the merge conflict and wait for tests to pass

**Pull request closed without merging**
